packages/backend/server/src/fundamentals/mutex/index.ts

```diff
@@ -0,0 +1,87 @@
+import { randomUUID } from 'node:crypto';
+
+import {
+  ForbiddenException,
+  Global,
+  Injectable,
+  Logger,
+  Module,
+} from '@nestjs/common';
+import { ClsService } from 'nestjs-cls';
+
+import { sleep } from '../utils/utils';
+
+export const MUTEX_RETRY = 3;
+export const MUTEX_WAIT = 100;
+
+@Injectable()
+export class MutexService {
+  private readonly logger = new Logger(MutexService.name);
+  private readonly bucket = new Map<string, string>();
+
+  constructor(private readonly cls: ClsService) {}
+
+  private getId() {
+    let id = this.cls.get('asyncId');
+
+    if (!id) {
+      id = randomUUID();
+      this.cls.set('asyncId', id);
+    }
+
+    return id;
+  }
+
+  async lockWith<R>(key: string, cb: () => Promise<R>): Promise<R> {
+    const locked = await this.lock(key);
+    if (locked) {
+      let result: R;
+      try {
+        result = await cb();
+      } finally {
+        await this.unlock(key);
+      }
+      return result;
+    } else {
+      throw new ForbiddenException('Failed to acquire lock');
```

Comment:
This error would be confusing because it will be directly shown to user. 
What if we pass lock fail error as parameter to callback just in the old school way, or you have a better idea?